### PR TITLE
Adding keccak256

### PR DESCRIPTION
**Pull request by Jim8y**

### CODE REVIEW

**Review by shargon — APPROVED** [review]

**Review by shargon — DISMISSED** [review]

**Review by vncoelho — CHANGES_REQUESTED** [review]
Decide about the possible attack as on 3
6.0 upgrade and also verify more tests.

**Comment by Liaojinghui** [comment]
> Decide about the possible attack as on 3.6.0 upgrade, and also verify more tests.

Hi vitor, what this means?

**Comment by Liaojinghui** [comment]
@shargon Tests ok and ready for merge.

**Review by shargon — APPROVED** [review]

**Comment by shargon** [comment]
> Decide about the possible attack as on 3.6.0 upgrade, and also verify more tests.

@vncoelho what this means?

**Comment by vncoelho** [comment]
Nowadays, we don't have this syscall,right?
What happens if someone calls it?
I am imagining that TX will fail.
Later, if this syscall exist the tx can be successful. It just need an if strategy to do malicious things.

These are the ForkTriggers @vang1ong7ang tried to explain.

I did test yet, but we can try that.

**Comment by Liaojinghui** [comment]
previous transactions can not call this cause this not yet exists, only if this is added will developers able to call it, since this will not affect previous transactions at all, no fork will introduced here.

**Comment by Liaojinghui** [comment]
@vncoelho Oh, I see what you mean now. But i dont think this should be an issue a pr to handle, otherwise, we will have a hardfork no matter what new functions we introduce to the core. update the consensus message maybe.

**Comment by vncoelho** [comment]
That is why this needs to be solved before this PR is merged. I did not reject the PR for nothing, there is a reason this should not be merged.

First, solve, then add new features.
Or, create a fork tag.

Otherwise, be prepared because at some point some attack will come...The attack is well known nowadays.

**Comment by Liaojinghui** [comment]
versioning things, consensus version, vm version all matters,,,,,,even plugin versions.........

**Comment by Liaojinghui** [comment]
@shargon hey grand master, are we using the `version` field of the block header? can we use it as the neo version?

**Comment by Liaojinghui** [comment]
will reopen after a proper solution to the version is found

**Comment by Liaojinghui** [comment]
@vncoelho man, you got any idea alrady? looks like this issue  is kinda complex since block header does not contain a stateroot

**Comment by vncoelho** [comment]
We have discussed some possibilities. I will create an issue starting this discussion.
A first idea would be to create an upgrade function for the native constructions. The calls would be related to contract versions.
The version can be added on the consensus phase.
All previous versions could be set to a default value.
But we are not yet sure it solves all cases. Lets discuss this on another issue.

**Comment by vang1ong7ang** [comment]
@Liaojinghui @shargon 

it is already discussed in #2910 : add-only changes also cause fork

> I'll also add that the scenario of "submitting a transaction with script FORKTRIGGER(); NEO.transfer(self, victim, amount) inside when v3.5.0, the attacker will get his money back when v3.6.0." is not always possible. If some PR is only adding some new native contract API then any attempt to call it will make transaction FAIL miserably. This means that after the upgrade it can do something instead of failing, but it's a bit harder to leverage that (not impossible though).

_Originally posted by @roman-khimov in https://github.com/neo-project/neo/issues/2910#issuecomment-1722137913_

> > I'll also add that the scenario of "submitting a transaction with script FORKTRIGGER(); NEO.transfer(self, victim, amount) inside when v3.5.0, the attacker will get his money back when v3.6.0." is not always possible.
> 
> They are basically the same. You can send a transaction first, and then use the second to check it's state in `FORKTRIGGER()`.
> 
> Any change including `add`, `delete` and `modify` should all be treated seriously.

_Originally posted by @dusmart in https://github.com/neo-project/neo/issues/2910#issuecomment-1722138811_

**Comment by Liaojinghui** [comment]
I really don't want to use the term fork here for dbft cause it will surly not cause a fork.....i would say it's a state inconsistency, for at most one block, the whole network will move on in the correct version. Though this also beaks the finality.

**Comment by Liaojinghui** [comment]
not suitable before we have a proper update mechanism for native contracts, close

**Comment by shargon** [comment]
> not suitable before we have a proper update mechanism for native contracts, close

I don't understand why until now it hasn't been a problem, and now suddenly new features cannot be added, this problem has been known since day one, it was decided to use the forks and check if it was compatible on the mainnet, and now without any debate, it is different.

**Comment by vncoelho** [comment]
@shargon, if it is known issue, why did you approve it without the fork tag?

**Comment by vncoelho** [comment]
@shargon, really, I am not understanding your behavior, specially because you have a strong background in security.
You are saying you known the issue. Why I need to come here and explain multiple times and reject the PR?

You were AGAIN, almost merging the PR that putting mainnet assets AT RISK!

I am here not saying that another strategy need to really be done before merging the PR. I said it is not ready to be merged. I did not recommended it to be closed.
But you should be more careful and consistent.

**Comment by shargon** [comment]
> @shargon, if it is known issue, why did you approve it without the fork tag?
> 

Because we are doing the same during 5 years

**Comment by vncoelho** [comment]
> Because we are doing the same during 5 years

@shargon, this is no excuse.
Let's end this because it is not benefit for any of us.

The issue is now well-known and was made public recently. So, let's, at least, be sure that every PR handles everything safely before merging. It can be with a new feature, using NativeUpdateHistory or a forktag.

**Comment by shargon** [comment]
> The issue is now well-known and was made public **recently**.

2019 - 
https://github.com/neo-project/neo/issues/1287#issue-528843931, also you can read the https://github.com/neo-project/neo-vm/issues/142 thread

**Comment by shargon** [comment]
@Liaojinghui could you rebase it with `update-native-contracts` and create a hardfork for it in order to test this feature?

**Comment by vncoelho** [comment]
you mean rebase to https://github.com/neo-project/neo/pull/2942#pullrequestreview-1714839134?

**Comment by shargon** [comment]
> you mean rebase to [#2942 (review)](https://github.com/neo-project/neo/pull/2942#pullrequestreview-1714839134)?

Yes

File: src/Neo/SmartContract/Native/CryptoLib.cs [hunk 1]
```diff
@@ -63,6 +65,21 @@ public static byte[] Murmur32(byte[] data, uint seed)
             return murmur.ComputeHash(data);
         }
 
+        /// <summary>
+        /// Computes the hash value for the specified byte array using the keccak256 algorithm.
+        /// </summary>
+        /// <param name="data">The input to compute the hash code for.</param>
+        /// <returns>Computed hash</returns>
+        [ContractMethod(CpuFee = 1 << 15, ActiveIn = Hardfork.HF_Manticore)]
```
Comment (shargon) [inline src/Neo/SmartContract/Native/CryptoLib.cs:73]: @roman-khimov @Liaojinghui nullable Enums are not valid as param in attributes... Back to HF_Genesis?

File: src/Neo/Hardfork.cs [hunk 2]
```diff
@@ -13,6 +13,7 @@ namespace Neo
     public enum Hardfork : byte
     {
         HF_Aspidochelone,
-        HF_Basilisk
+        HF_Basilisk,
+        HF_Manticore
```
Comment (roman-khimov) [inline src/Neo/Hardfork.cs:17]: Can't be.

File: src/Neo/SmartContract/Native/CryptoLib.cs (hunk 1 above)
Comment (roman-khimov) [reply src/Neo/SmartContract/Native/CryptoLib.cs]: Too C#-specific for me to handle, but in general I still don't think we need a `Genesis` hardfork and changing constants.

**Review by roman-khimov — COMMENTED** [review]
Let's concentrate on #2914 first, this thing deserves more attention and then we'll deal with protocol extensions.

File: src/Neo/Hardfork.cs (hunk 2 above)
Comment (AnnaShaleva) [reply src/Neo/Hardfork.cs:17]: Centaurus!

Comment (roman-khimov) [reply src/Neo/Hardfork.cs:17]: https://github.com/neo-project/neo/pull/2749#discussion_r878971641

Sorry.

**Comment by shargon** [comment]
> Let's concentrate on #2914 first, this thing deserves more attention and then we'll deal with protocol extensions.

It's for testing, it will be easier with a valid one case

File: src/Neo/Hardfork.cs (hunk 2 above)
Comment (shargon) [reply src/Neo/Hardfork.cs:17]: Chat-Gpt said `Manticore` xD

Comment (shargon) [reply src/Neo/Hardfork.cs:18]:
```suggestion
        HF_Cockatrice
```

File: src/Neo/SmartContract/Native/CryptoLib.cs [hunk 3]
```diff
@@ -64,6 +66,21 @@ public static byte[] Murmur32(byte[] data, uint seed)
             return murmur.ComputeHash(data);
         }
 
+        /// <summary>
+        /// Computes the hash value for the specified byte array using the keccak256 algorithm.
+        /// </summary>
+        /// <param name="data">The input to compute the hash code for.</param>
+        /// <returns>Computed hash</returns>
+        [ContractMethod(Hardfork.HF_Manticore, CpuFee = 1 << 15)]
```
Comment (shargon) [inline src/Neo/SmartContract/Native/CryptoLib.cs:74]:
```suggestion
        [ContractMethod(Hardfork.HF_Cockatrice, CpuFee = 1 << 15)]
```

**Comment by shargon** [comment]
Require https://github.com/neo-project/neo/pull/2942

**Review by shargon — APPROVED** [review]

**Comment by shargon** [comment]
It will be merged in `update-native-contract` brach

**Comment by Jim8y** [comment]
@AnnaShaleva @roman-khimov  neo-go will update accordingly.

**Comment by AnnaShaleva** [comment]
Sure, will be implemented together with native contracts update.

**Pull request merged**